src/Microsoft.Diagnostics.Monitoring/DiagnosticsEventPipeProcessor.cs

```diff
@@ -289,6 +317,96 @@ private void PostMetric(Metric metric)
             }
         }
 
+        private Task HandleGCEvents(EventPipeEventSource source, int pid, Func<Task> stopFunc, CancellationToken token)
+        {
+            int gcNum = -1;
+
+            TaskCompletionSource<object> gcEventsStarted = new TaskCompletionSource<object>();
+            token.Register(() => gcEventsStarted.TrySetCanceled(token));
+
+            TaskCompletionSource<object> gcStopped = new TaskCompletionSource<object>();
+            token.Register(() => gcStopped.TrySetCanceled(token));
+
+            source.Clr.GCStart += data =>
+            {
+                if (data.ProcessID != pid)
+                {
+                    return;
+                }
+
+                gcEventsStarted.TrySetResult(null);
+
+                if (gcNum < 0 && data.Depth == 2 && data.Type != GCType.BackgroundGC)
+                {
+                    gcNum = data.Count;
+                }
+            };
+
+            source.Clr.GCStop += data =>
+            {
+                if (data.ProcessID != pid)
+                {
+                    return;
+                }
+
+                if (data.Count == gcNum)
+                {
+                    gcStopped.TrySetResult(null);
+                }
+            };
+
+            source.Clr.GCBulkNode += data =>
+            {
+                if (data.ProcessID != pid)
+                {
+                    return;
+                }
+
+                gcEventsStarted.TrySetResult(null);
+            };
+
+            DotNetHeapDumpGraphReader dumper = new DotNetHeapDumpGraphReader(TextWriter.Null)
+            {
+                DotNetHeapInfo = new DotNetHeapInfo()
+            };
+            dumper.SetupCallbacks(_gcGraph, source, pid.ToString(CultureInfo.InvariantCulture));
+
+            return Task.Run(async () =>
+            {
+                // Wait for GC events for up to 5 seconds.
+                Task eventsStartedTask = gcEventsStarted.Task;
+                Task eventsTimeoutTask = Task.Delay(TimeSpan.FromSeconds(5), token);
```

Comment:
Why is this hardcoded to 5 seconds?

Reply:
This is what the existing dotnet-gcdump tool does. I don't see much in terms of comments around why it's doing that. One of the conditions that can be met before attempting to serialize the heap information is to stop waiting for the events if no gc events are received within 5 seconds (see https://github.com/dotnet/diagnostics/blob/bf96a0a7c86850758e5ba5065b916449356b43ab/src/Tools/dotnet-gcdump/DotNetHeapDump/EventPipeDotNetHeapDumper.cs#L157).

What I've observed why trying to figure out why it does this is that the events are not always provided by the event source freely until the event source is told to stop processing events. Then it will flush any events that it got, thus the GC event callbacks are raised. So I think the existing dotnet-gcdump tool is given the event source some small amount of time to collect the events (even if they aren't published) before forcing it to stop processing and flush the events.

I'll add comments to this new implementation to describe what is happening.

Reply:
This is a hold over from PerfView that has the same timeout.  The logic being invoked here is a clone of code from the PerfView repo to keep the gcdumps consistent.  I believe Justin is correct in his assumption about event publishing, but I never did enough research into why that happens to say so definitively.